File: algorithm/Algorithm/Code_day04_BuildingOutline.java

package zuoshen.up.advanced_class_04;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Map.Entry;
import java.util.TreeMap;

public class Code_01_Building_Outline {

    public static class Node {
         public boolean isUp;
         public int posi;
         public int h;
 
         public Node(boolean bORe, int position, int height) {
        isUp = bORe;
        posi = position;
        h = height;
    }
}

public static class NodeComparator implements Comparator<Node> {
    @Override
    public int compare(Node o1, Node o2) {
        if (o1.posi != o2.posi) {
            return o1.posi - o2.posi;
        }
        if (o1.isUp != o2.isUp) {
            return o1.isUp ? -1 : 1;
        }
        return 0;
    }
}

/**
 * 大楼轮廓线
 * 如果已经知道了每个idx上的最大高度，就很容求出来最后的轮廓线
 * 
 * 所以首先要求出每个位置上的最大高度，也就是这里的pmMap(idx，最大高度)
 * 求法：
 * 每个大楼有2个边界，分别是上/下
 * 产生边界数组之后，按照pos进行升序排序，位置相同那么先up后down
 * 排序完毕，遍历产生的边界数组，用treemap记录(高度，高度出现的次数)，up则次数+1，down则次数-1
 * 同时，遍历的时候，记录边界数组的pos和treemap中的最大key，也就是pos和pos上的最大高度=>形成(pos,MaxHeight)的map
 * 之后，遍历map，如果最大高度发生变化，那么形成一次轮廓线
 *      高度不发生变化，进入下一次循环，直到遍历完毕
 *      另外curHeight=0说明是高楼的左轮廓线
 */
	public static List<List<Integer>> buildingOutline(int[][] buildings) {
		Node[] nodes = new Node[buildings.length * 2];
		for (int i = 0; i < buildings.length; i++) {
			nodes[i * 2] = new Node(true, buildings[i][0], buildings[i][2]);
			nodes[i * 2 + 1] = new Node(false, buildings[i][1], buildings[i][2]);
		}
		Arrays.sort(nodes, new NodeComparator());
		TreeMap<Integer, Integer> htMap = new TreeMap<>();
		TreeMap<Integer, Integer> pmMap = new TreeMap<>();
		for (int i = 0; i < nodes.length; i++) {
			if (nodes[i].isUp) {
				if (!htMap.containsKey(nodes[i].h)) {
					htMap.put(nodes[i].h, 1);
				} else {
					htMap.put(nodes[i].h, htMap.get(nodes[i].h) + 1);
				}
			} else {
				if (htMap.containsKey(nodes[i].h)) {
					if (htMap.get(nodes[i].h) == 1) {
						htMap.remove(nodes[i].h);
					} else {
						htMap.put(nodes[i].h, htMap.get(nodes[i].h) - 1);
					}
				}
			}
			if (htMap.isEmpty()) {
				pmMap.put(nodes[i].posi, 0);
			} else {
				pmMap.put(nodes[i].posi, htMap.lastKey());
			}
		}
		List<List<Integer>> res = new ArrayList<>();
		int start = 0;
		int height = 0;
		for (Entry<Integer, Integer> entry : pmMap.entrySet()) {
			int curPosition = entry.getKey();
			int curMaxHeight = entry.getValue();
			if (height != curMaxHeight) {
				if (height != 0) {
					List<Integer> newRecord = new ArrayList<Integer>();
					newRecord.add(start);
					newRecord.add(curPosition);
					newRecord.add(height);
					res.add(newRecord);
				}
				start = curPosition;
				height = curMaxHeight;
			}
		}
		return res;
	}

}

File: algorithm/Algorithm/Code_day04_BuildingOutline.js

class Node {
	constructor(pos, height, up) {
		this.pos = pos;
		this.height = height;
		this.up = up;
	}
}

function compare(a, b) {
	if (a.pos != b.pos) {
		return a.pos - b.pos;
	}
	if (a.up != b.up) {
		return !a.up;
	}
	return 0;
}

Map.prototype.LastKey = function () {
	let maxKey = 0;
	this.forEach((v, k) => {
		maxKey = Math.max(maxKey, k);
	})
	return maxKey;
}

function buildingOutline(buildings) {
	let mapHeightCount = new Map(); // (高度，次数)
	let nodes = new Array(buildings.length * 2);
	for (let i = 0; i < buildings.length; i++) {
		nodes[2 * i] = new Node(buildings[i][0], buildings[i][2], true);
		nodes[2 * i + 1] = new Node(buildings[i][1], buildings[i][2], false);
	}
	nodes.sort(compare);
	// console.log(nodes)
	let mapPosMaxheight = new Map();
	for (let i = 0; i < nodes.length; i++) {
		// 用来更新高度的次数
		if (nodes[i].up) {
			if (mapHeightCount.has(nodes[i].height)) {
				mapHeightCount.set(nodes[i].height, mapHeightCount.get(nodes[i].pos) + 1);
			} else {
				mapHeightCount.set(nodes[i].height, 1);
			}
		} else {
			if (mapHeightCount.has(nodes[i].height)) {
				if (mapHeightCount.get(nodes[i].height) == 1) {
					mapHeightCount.delete(nodes[i].height);
				} else {
					mapHeightCount.set(nodes[i].height, mapHeightCount.get(nodes[i].height) - 1);
				}
			}
		}
		// 更新完毕之后，记录每个pos的MaxHeight
		if (mapHeightCount.size == 0) {
			mapPosMaxheight.set(nodes[i].pos, 0);
		} else {
			mapPosMaxheight.set(nodes[i].pos, mapHeightCount.LastKey()); // JS的语言限制，没有TreeMap，这里仅仅是描述一下
		}
	}
	// 至此，得到了mapPosMaxheight，处理一下就能得到最后结果
	let res = [];
	let height = 0;
	let start = 0;
	mapPosMaxheight.forEach((curheight, curPos) => {
		if (height != curheight) {
			if (height != 0) {
				res.push([start, curPos, height])
			}
			start = curPos;
			height = curheight;
		}
	})
	return res;
}

// for test
function main() {
	console.log(buildingOutline([[1, 3, 3], [2, 4, 4], [5, 6, 1]]))
}

main();